src/Analyzers/CSharp/Analyzers/UsePatternCombinators/CSharpUsePatternCombinatorsAnalyzer.cs

```diff
@@ -0,0 +1,169 @@
+    using static BinaryOperatorKind;
+    using static AnalyzedPattern;
+
+    internal static class CSharpUsePatternCombinatorsAnalyzer
+    {
+        public static AnalyzedPattern? Analyze(IOperation operation)
+        {
+            var pattern = ParsePattern(operation);
+            return pattern?.Target.Syntax is ExpressionSyntax ? pattern : null;
+        }
+
+        private enum ConstantResult
+        {
+            /// <summary>
+            /// None of operands were constant.
+            /// </summary>
+            None,
+
+            /// <summary>
+            /// The left operand is the constant.
+            /// </summary>
+            Left,
+
+            /// <summary>
+            /// The right operand is the constant.
+            /// </summary>
+            Right,
+        }
+
+        private static AnalyzedPattern? ParsePattern(IOperation operation)
+        {
+            switch (operation)
+            {
+                case IBinaryOperation { OperatorKind: BinaryOperatorKind.Equals } op:
+                    return ParseConstantPattern(op);
+
+                case IBinaryOperation { OperatorKind: NotEquals } op:
+                    return Not.Create(ParseConstantPattern(op));
+
+                case IBinaryOperation { OperatorKind: ConditionalOr, Syntax: BinaryExpressionSyntax syntax } op:
+                    return ParseBinaryPattern(op, isDisjunctive: true, syntax.OperatorToken);
+
+                case IBinaryOperation { OperatorKind: ConditionalAnd, Syntax: BinaryExpressionSyntax syntax } op:
+                    return ParseBinaryPattern(op, isDisjunctive: false, syntax.OperatorToken);
+
+                case IBinaryOperation op when IsRelationalOperator(op.OperatorKind):
+                    return ParseRelationalPattern(op);
+
+                case IUnaryOperation { OperatorKind: UnaryOperatorKind.Not } op:
+                    return Not.Create(ParsePattern(op.Operand));
+
+                case IIsTypeOperation { Syntax: BinaryExpressionSyntax { Right: TypeSyntax type } } op:
+                    return new Type(type, op.ValueOperand);
+
+                case IIsPatternOperation { Pattern: { Syntax: PatternSyntax pattern } } op:
+                    return new Source(pattern, op.Value);
+
+                case IParenthesizedOperation op:
+                    return ParsePattern(op.Operand);
+            }
+
+            return null;
+        }
+
+        private static AnalyzedPattern? ParseBinaryPattern(IBinaryOperation op, bool isDisjunctive, SyntaxToken token)
+        {
+            var leftPattern = ParsePattern(op.LeftOperand);
+            if (leftPattern == null)
+                return null;
+
+            var rightPattern = ParsePattern(op.RightOperand);
+            if (rightPattern == null)
+                return null;
+
+            return Binary.Create(leftPattern, rightPattern, isDisjunctive, token);
+        }
+
+        private static ConstantResult DetermineConstant(IBinaryOperation op)
```

Comment:
nit, these methods could be =>

Reply:
I'm not really comfortable using (multiline) switch expressions with expr-bodied methods. It's just doesn't look any cleaner? maybe I should change others to be consistent

Reply:
no worries.  that's what 'nits' are :)